pgmax/fg/graph.py

```diff
@@ -216,15 +260,32 @@ def log_potentials(self) -> np.ndarray:
             A jnp array representing the log of the potential function for each
                 valid configuration
         """
+        if not np.any(
+            [
+                isinstance(
+                    factor_group,
+                    (groups.EnumerationFactorGroup, groups.PairwiseFactorGroup),
+                )
+                for factor_group in self.factor_groups
+            ]
+        ):
+            return None
+
         return np.concatenate(
             [
                 factor_group.factor_group_log_potentials
                 for factor_group in self.factor_groups
+                if isinstance(
+                    factor_group,
+                    (groups.EnumerationFactorGroup, groups.PairwiseFactorGroup),
+                )
             ]
         )
 
     @cached_property
-    def factors(self) -> Tuple[nodes.EnumerationFactor, ...]:
+    def factors(
```

Comment:
Related to the above comments, instead of returning a single tuple we should return a dictionary, with one element per factor type.

Reply:
Not here: we need a flattened list of factors because we will have a flattened message to factors

Reply:
I don't think these are contradictory. We basically cut the flat messages into a few sections, one per factor type. Each factor type has access to start and end indices and does message updates using only that section of the messages